libcfnet/file_stream.c

```diff
@@ -0,0 +1,943 @@
+        }
+    } while (res != RS_DONE);
+
+    fclose(file);
+    rs_job_free(job);
+
+    return true;
+}
+
+bool FileStreamServe(SSL *conn, const char *filename)
+{
+    assert(conn != NULL);
+    assert(filename != NULL);
+
+    Log(LOG_LEVEL_VERBOSE,
+        "Receiving- & loading signature into memory for file '%s'...",
+        filename);
+    rs_signature_t *sig;
+    if (!RecvSignature(conn, &sig))
+    {
+        /* Error is already logged */
+        return false;
+    }
+
+    Log(LOG_LEVEL_VERBOSE,
+        "Computing- & sending delta for file '%s'...",
+        filename);
+    if (!SendDelta(conn, sig, filename))
+    {
+        /* Error is already logged */
+        rs_free_sumset(sig);
+        return false;
+    }
+
+    rs_free_sumset(sig);
+    return true;
+}
+
+/*********************************************************/
+/* Client specific                                       */
+/*********************************************************/
+
+#define ERROR_MSG_INTERNAL_CLIENT_ERROR "Internal client error"
+
+/**
+ * @brief Compute and send a signature of the basis file
+ *
+ * @param conn The SSL connection object
+ * @param filename The name of the basis file
+ * @return true on success, otherwise false
+ */
+static bool SendSignature(SSL *conn, const char *filename)
+{
+    assert(conn != NULL);
+    assert(filename != NULL);
+
+    /* In this case, the input buffer does not need to be twice the message
+     * size, because we can control how much we read into it */
+    char in_buf[MESSAGE_SIZE], out_buf[MESSAGE_SIZE];
+
+    /* Open basis file */
+    FILE *file = safe_fopen(filename, "rb");
+    if (file == NULL)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to open the basis file '%s' for computing delta during file stream: %s",
+            filename,
+            GetErrorStr());
+        SendError(conn, false, ERROR_MSG_INTERNAL_CLIENT_ERROR);
+        return false;
+    }
+
+    /* Get file size */
+    rs_long_t fsize = rs_file_size(file);
+
+    /* Get recommended arguments */
+    rs_magic_number sig_magic = 0;
+    size_t block_len = 0, strong_len = 0;
+    rs_result res = rs_sig_args(fsize, &sig_magic, &block_len, &strong_len);
+    if (res != RS_DONE)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to get recommended signature arguments: %s",
+            rs_strerror(res));
+        SendError(conn, false, ERROR_MSG_INTERNAL_CLIENT_ERROR);
+        fclose(file);
+        return false;
+    }
+
+    /* Start generating signature */
+    rs_job_t *job = rs_sig_begin(block_len, strong_len, sig_magic);
+    if (job == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Failed to begin job for generating signature");
+        SendError(conn, false, ERROR_MSG_INTERNAL_CLIENT_ERROR);
+        fclose(file);
+        return false;
+    }
+
+    /* Setup buffers */
+    rs_buffers_t bufs = {0};
+    bufs.next_out = out_buf;
+    bufs.avail_out = MESSAGE_SIZE; /* We cannot send more using the protocol */
+
+    do
```

Comment:
Any chance we could de-duplicate what's in this `do-while` with the one in `SendDelta()`? It's very long anyway...

Reply:
Yes, I think it should be possible. The log messages would have to be more generic and the code would probably be a bit harder to understand. I will create a follow-up ticket to investigate this further.

Reply:
https://northerntech.atlassian.net/browse/CFE-4456